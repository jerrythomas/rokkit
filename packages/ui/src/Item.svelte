<script>
	import Icon from './Icon.svelte'
	import { FieldMapper } from '@rokkit/core'

	/**
	 * @typedef {Object} Props
	 * @property {string|Object} [value]
	 * @property {FieldMapper}   [mapping]
	 */

	/** @type {Props} */
	let { value, mapping = new FieldMapper() } = $props()
	let content = $derived(mapping.getText(value))
	let ariaLabel = $derived(mapping.getLabel(value) ?? content)
</script>

{#if mapping.hasIcon(value)}
	<Icon name={mapping.getIcon(value)} label={ariaLabel} />
{:else if mapping.hasImage(value)}
	<img src={mapping.getImage(value)} alt={ariaLabel} />
{/if}
{#if content}
	<p>{content}</p>
{/if}
