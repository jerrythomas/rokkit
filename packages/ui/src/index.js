// skipcq: JS-E1004 - Needed for exposing all components
export * from '@rokkit/atoms'
// skipcq: JS-E1004 - Needed for exposing all components
export * from '@rokkit/input'
