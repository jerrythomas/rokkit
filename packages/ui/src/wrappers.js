// skipcq: JS-E1004 - Needed for exposing all components
export * from '@rokkit/organisms/wrappers'
