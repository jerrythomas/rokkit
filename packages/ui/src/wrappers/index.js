export { default as Wrapper } from './Wrapper.svelte'
export { default as Section } from './Section.svelte'
export { default as Category } from './Category.svelte'
