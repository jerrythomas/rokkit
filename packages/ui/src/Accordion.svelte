<script>
	import { equals } from 'ramda'
	import { createEmitter, noop, getKeyFromPath, getSnippet } from '@rokkit/core'
	import { NestedController } from '@rokkit/states'
	import { navigator } from '@rokkit/actions'
	import Summary from './Summary.svelte'
	import Item from './Item.svelte'
	import ListBody from './ListBody.svelte'

	/**
	 * @typedef {Object} Props
	 * @property {string} [class]
	 * @property {any} [items]
	 * @property {import('@rokkit/core').FieldMapper} [mapping]
	 * @property {boolean} [autoCloseSiblings]
	 * @property {boolean} [multiselect]
	 * @property {any} [value]
	 */

	/** @type {Props} */
	let {
		class: classes = '',
		items = $bindable([]),
		value = $bindable(null),
		fields,
		autoCloseSiblings = false,
		multiselect = false,
		header = null,
		footer = null,
		empty = null,
		stub = null,
		extra,
		...events
	} = $props()

	let emitter = $derived(
		createEmitter(events, ['collapse', 'change', 'expand', 'click', 'select', 'move'])
	)
	let wrapper = new NestedController(items, value, fields, {
		events,
		multiselect,
		autoCloseSiblings
	})
</script>

<!-- svelte-ignore a11y_no_noninteractive_tabindex -->
<rk-accordion
	class={classes}
	tabindex="0"
	use:navigator={{ wrapper }}
	onactivate={() => (value = wrapper.value)}
>
	{#if header}
		<rk-header>{@render header()}</rk-header>
	{/if}
	{#if items.length === 0}
		<rk-list-item role="presentation">
			{#if empty}
				{@render empty()}
			{:else}
				No items found.
			{/if}
		</rk-list-item>
	{/if}
	{#each items as item, index}
		{@const key = `${index}`}
		<div
			class="flex flex-col"
			class:is-expanded={wrapper.expandedKeys.has(key)}
			class:is-selected={wrapper.selectedKeys.has(key)}
			data-path={index}
		>
			<Summary
				bind:value={items[index]}
				{fields}
				expanded={wrapper.expandedKeys.has(key)}
				hasChildren={wrapper.hasChildren(key)}
			/>
			{#if wrapper.expandedKeys.has(key)}
				<rk-list role="listbox" tabindex="-1">
					<ListBody
						bind:items={items[fields.children]}
						bind:value
						fields={fields.fields ?? fields}
						{selected}
						{stub}
						onchange={emitter.change}
						{extra}
					/>
				</rk-list>
			{/if}
		</div>
	{/each}
	{#if footer}
		<rk-footer>{@render footer()}</rk-footer>
	{/if}
</rk-accordion>
