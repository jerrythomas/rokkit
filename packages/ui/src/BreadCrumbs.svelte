<script>
	import { getSnippet } from '@rokkit/core'
	import Item from './Item.svelte'

	/**
	 * @typedef {Object} Props
	 * @property {string} [class]
	 * @property {any} [items]
	 * @property {string} [separator]
	 * @property {any} [fields]
	 * @property {any} [crumb]
	 */

	/** @type {Props} */
	let { class: classes = '', items = [], separator = '/', fields, crumb } = $props()
</script>

<rk-crumbs class={classes}>
	{#each items as item, index (index)}
		{#if index > 0}
			<span>
				{#if separator.length === 1}
					{separator}
				{:else}
					<icon class={separator}></icon>
				{/if}
			</span>
		{/if}
		<rk-crumb class:is-selected={index === items.length - 1}>
			{#if crumb}
				{@render crumb(item, fields)}
			{:else}
				<Item value={item} {fields} />
			{/if}
		</rk-crumb>
	{/each}
</rk-crumbs>
