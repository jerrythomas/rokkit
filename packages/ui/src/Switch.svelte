<script>
	import { equals } from 'ramda'
	import { noop, FieldMapper } from '@rokkit/core'
	import { keyboard } from '@rokkit/actions'
	import Item from './Item.svelte'
	// import { defaultMapping } from './constants'

	/**
	 * @typedef {Object} Props
	 * @property {string}      [class]
	 * @property {any}         value
	 * @property {Array<any>}  [options]
	 * @property {FieldMapper} [mapping]
	 * @property {boolean}     [compact]
	 * @property {boolean}     [disabled]
	 */

	/** @type {Props} */
	let {
		class: classes = '',
		value = $bindable(),
		options = [false, true],
		fields,
		compact = false,
		disabled = false,
		onchange = noop,
		stub,
		...extra
	} = $props()

	// let cursor = $state([])

	function toggle(direction = 1) {
		let nextIndex
		const index = options.indexOf(value)
		if (index === -1) {
			nextIndex = direction === -1 ? options.length - 1 : 0
		} else {
			nextIndex = (index + direction + options.length) % options.length
		}

		value = options[nextIndex]
		onchange(value)
	}

	function handleClick(event) {
		const index = event.target.closest('[data-path]').dataset.path
		value = options[index]
		onchange(value)
	}
	const keyMapping = {
		next: ['ArrowRight', 'ArrowDown', ' ', 'Enter'],
		prev: ['ArrowLeft', 'ArrowUp']
	}
	// let useComponent = $derived(!options.every((item) => [false, true].includes(item)))
	// let mapper = new FieldMapper(fields)
</script>

{#if !Array.isArray(options) || options.length < 2}
	<rk-error>Items should be an array with at least two items.</rk-error>
{:else}
	<!-- svelte-ignore a11y_click_events_have_key_events -->
	<rk-switch
		class={classes}
		class:is-off={options.length === 2 && equals(value, options[0])}
		class:is-on={options.length === 2 && equals(value, options[1])}
		class:compact
		aria-label="Toggle Switch"
		aria-orientation="horizontal"
		aria-disabled={disabled}
		tabindex="0"
		role="listbox"
		use:keyboard={keyMapping}
		onnext={() => toggle()}
		onprev={() => toggle(-1)}
		onclick={handleClick}
	>
		{#each options as item, index (item)}
			<!-- {@const Template = getSnippet(extra, mapper.get('snippet', item), stub)} -->
			<rk-item class="relative" role="option" aria-selected={equals(item, value)} data-path={index}>
				{#if equals(item, value)}
					<rk-indicator class="absolute bottom-0 left-0 right-0 top-0"></rk-indicator>
				{/if}
				{#if stub}
					{@render stub(item, fields)}
					<!-- <Template value={item} {fields} /> -->
				{:else}
					<Item value={item} {fields} />
				{/if}
			</rk-item>
		{/each}
	</rk-switch>
{/if}
