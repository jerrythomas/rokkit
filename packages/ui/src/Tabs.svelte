<script>
	import { defaultFields, defaultStateIcons, noop } from '@rokkit/core'
	import { defaultMapping } from './constants'
	import { navigator } from '@rokkit/actions'
	import Icon from './Icon.svelte'
	import Item from './Item.svelte'

	/**
	 * @typedef {Object} Props
	 * @property {string} [class]
	 * @property {any} [options]
	 * @property {import('@rokkit/core').FieldMapping} [mapping]
	 * @property {any} [value]
	 * @property {boolean} [below]
	 * @property {string} [align]
	 * @property {boolean} [editable]
	 * @property {any} [icons]
	 */

	/** @type {Props} */
	let {
		class: className = '',
		options = $bindable([]),
		fields = defaultFields,
		mapping = defaultMapping,
		value = $bindable(null),
		below = false,
		align = 'left',
		editable = false,
		icons = $bindable(defaultStateIcons.action),
		onremove = noop,
		onadd = noop,
		onselect = noop
	} = $props()

	let cursor = $state([])

	function handleRemove(event) {
		if (typeof event.detail === Object) {
			event.detail[mapping.fields.isDeleted] = true
		} else {
			options = options.filter((i) => i !== event.detail)
		}

		onremove({ item: event.detail })
	}
	function handleAdd(event) {
		event.stopPropagation()
		onadd()
	}
	function handleNav(event) {
		value = event.detail.node
		cursor = event.detail.path

		onselect({ item: value, indices: cursor })
	}
	let stateIcons = $derived({ ...defaultStateIcons.action, ...icons })
	let filtered = $derived(options.filter((item) => !item[fields.isDeleted]))
</script>

<rk-tabs
	class="flex w-full {className}"
	class:is-below={below}
	class:justify-center={align === 'center'}
	class:justify-end={align === 'right'}
	tabindex="0"
	role="listbox"
	onmove={handleNav}
	onselect={handleNav}
>
	{#each filtered as item, index}
		{@const Template = mapping.getComponent(item)}
		<rk-tab>
			<Template value={item} {mapping} />
			{#if editable}
				<Icon
					name="remove"
					role="button"
					label="Delete Tab"
					size="small"
					onclick={() => handleRemove(item)}
				/>
			{/if}
		</rk-tab>
	{/each}
	{#if editable}
		<Icon name="add" role="button" label="Add Tab" size="small" on:click={handleAdd} />
	{/if}
</rk-tabs>
