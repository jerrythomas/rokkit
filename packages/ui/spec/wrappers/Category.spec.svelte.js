import { describe, expect, it, beforeEach } from 'vitest'
import { cleanup, render } from '@testing-library/svelte'
import { flushSync, tick } from 'svelte'
import { Wrapper, Category } from '../../src/wrappers'
import Switch from '../../src/Switch.svelte'
import Item from '../../src/Item.svelte'
import Tabs from '../../src/Tabs.svelte'

describe('Category.svelte', () => {
	const items = ['one', 'two', 'three']
	const using = {
		editors: {},
		components: {},
		wrappers: {},
		navigators: {}
	}
	const registry = $state({
		editors: { ...using.editors },
		components: { default: Item, ...using.components },
		wrappers: { default: Wrapper, ...using.wrappers },
		navigators: { default: Tabs, ...using.navigators }
	})

	beforeEach(() => {
		cleanup()
	})

	it('should render using default props', () => {
		const props = $state({ options: items })
		const { container } = render(Category, {
			context: new Map([['registry', registry]]),
			props
		})
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()
	})

	it('should render using specified type', () => {
		const props = $state({ options: items, type: 'horizontal' })
		const { container } = render(Category, {
			context: new Map([['registry', registry]]),
			props
		})
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()

		props.type = 'section'
		flushSync()
		expect(container).toMatchSnapshot()
	})

	it('should render using specified class', () => {
		const props = $state({ options: items, class: 'custom-class' })
		const { container } = render(Category, {
			context: new Map([['registry', registry]]),
			props
		})
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()

		props.class = 'custom-class-2'
		flushSync()
		expect(container).toMatchSnapshot()
	})
	it('should render with extra props', async () => {
		const props = $state({ options: items, align: 'center' })
		const { container } = render(Category, {
			context: new Map([['registry', registry]]),
			props
		})
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()

		props.align = 'right'
		flushSync()
		expect(container).toMatchSnapshot()
	})
	it('should render with alternative navigator', async () => {
		registry.navigators = { ...registry.navigators, Switch }
		const props = $state({ options: items, navigator: 'Switch' })
		const { container } = render(Category, {
			context: new Map([['registry', registry]]),
			props
		})
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()

		props.navigator = 'tabs'
		flushSync()
		expect(container).toMatchSnapshot()
	})
})
