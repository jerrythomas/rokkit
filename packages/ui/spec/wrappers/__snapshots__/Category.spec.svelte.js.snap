// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Category.svelte > should render using default props 1`] = `
<div>
  <section>
    <!---->
    <rk-tabs
      class="flex w-full "
      role="listbox"
      tabindex="0"
    >
      <rk-tab>
        <!---->
        <!---->
         
        <p>
          one
        </p>
        <!---->
        <!---->
         
        <!---->
      </rk-tab>
      <rk-tab>
        <!---->
        <!---->
         
        <p>
          two
        </p>
        <!---->
        <!---->
         
        <!---->
      </rk-tab>
      <rk-tab>
        <!---->
        <!---->
         
        <p>
          three
        </p>
        <!---->
        <!---->
         
        <!---->
      </rk-tab>
      <!---->
       
      <!---->
    </rk-tabs>
    
    <!---->
     
    <field-layout
      class="vertical"
    >
      <!---->
    </field-layout>
  </section>
  
</div>
`;

exports[`Category.svelte > should render using specified class 1`] = `
<div>
  <section
    class="custom-class"
  >
    <!---->
    <rk-tabs
      class="flex w-full "
      role="listbox"
      tabindex="0"
    >
      <rk-tab>
        <!---->
        <!---->
         
        <p>
          one
        </p>
        <!---->
        <!---->
         
        <!---->
      </rk-tab>
      <rk-tab>
        <!---->
        <!---->
         
        <p>
          two
        </p>
        <!---->
        <!---->
         
        <!---->
      </rk-tab>
      <rk-tab>
        <!---->
        <!---->
         
        <p>
          three
        </p>
        <!---->
        <!---->
         
        <!---->
      </rk-tab>
      <!---->
       
      <!---->
    </rk-tabs>
    
    <!---->
     
    <field-layout
      class="vertical"
    >
      <!---->
    </field-layout>
  </section>
  
</div>
`;

exports[`Category.svelte > should render using specified class 2`] = `
<div>
  <section
    class="custom-class-2"
  >
    <!---->
    <rk-tabs
      class="flex w-full "
      role="listbox"
      tabindex="0"
    >
      <rk-tab>
        <!---->
        <!---->
         
        <p>
          one
        </p>
        <!---->
        <!---->
         
        <!---->
      </rk-tab>
      <rk-tab>
        <!---->
        <!---->
         
        <p>
          two
        </p>
        <!---->
        <!---->
         
        <!---->
      </rk-tab>
      <rk-tab>
        <!---->
        <!---->
         
        <p>
          three
        </p>
        <!---->
        <!---->
         
        <!---->
      </rk-tab>
      <!---->
       
      <!---->
    </rk-tabs>
    
    <!---->
     
    <field-layout
      class="vertical"
    >
      <!---->
    </field-layout>
  </section>
  
</div>
`;

exports[`Category.svelte > should render using specified type 1`] = `
<div>
  <section>
    <!---->
    <rk-tabs
      class="flex w-full "
      role="listbox"
      tabindex="0"
    >
      <rk-tab>
        <!---->
        <!---->
         
        <p>
          one
        </p>
        <!---->
        <!---->
         
        <!---->
      </rk-tab>
      <rk-tab>
        <!---->
        <!---->
         
        <p>
          two
        </p>
        <!---->
        <!---->
         
        <!---->
      </rk-tab>
      <rk-tab>
        <!---->
        <!---->
         
        <p>
          three
        </p>
        <!---->
        <!---->
         
        <!---->
      </rk-tab>
      <!---->
       
      <!---->
    </rk-tabs>
    
    <!---->
     
    <field-layout
      class="horizontal"
    >
      <!---->
    </field-layout>
  </section>
  
</div>
`;

exports[`Category.svelte > should render using specified type 2`] = `
<div>
  <section>
    <!---->
    <rk-tabs
      class="flex w-full "
      role="listbox"
      tabindex="0"
    >
      <rk-tab>
        <!---->
        <!---->
         
        <p>
          one
        </p>
        <!---->
        <!---->
         
        <!---->
      </rk-tab>
      <rk-tab>
        <!---->
        <!---->
         
        <p>
          two
        </p>
        <!---->
        <!---->
         
        <!---->
      </rk-tab>
      <rk-tab>
        <!---->
        <!---->
         
        <p>
          three
        </p>
        <!---->
        <!---->
         
        <!---->
      </rk-tab>
      <!---->
       
      <!---->
    </rk-tabs>
    
    <!---->
     
    <field-layout
      class="section"
    >
      <!---->
    </field-layout>
  </section>
  
</div>
`;

exports[`Category.svelte > should render with alternative navigator 1`] = `
<div>
  <section>
    <!---->
    <!---->
    <rk-switch
      aria-disabled="false"
      aria-label="Toggle Switch"
      aria-orientation="horizontal"
      role="listbox"
      tabindex="0"
    >
      <rk-item
        aria-selected="false"
        class="relative"
        data-path="0"
        role="option"
      >
        <!---->
         
        <!---->
      </rk-item>
      <rk-item
        aria-selected="false"
        class="relative"
        data-path="1"
        role="option"
      >
        <!---->
         
        <!---->
      </rk-item>
      <rk-item
        aria-selected="false"
        class="relative"
        data-path="2"
        role="option"
      >
        <!---->
         
        <!---->
      </rk-item>
      
    </rk-switch>
    
    
    <!---->
     
    <field-layout
      class="vertical"
    >
      <!---->
    </field-layout>
  </section>
  
</div>
`;

exports[`Category.svelte > should render with alternative navigator 2`] = `
<div>
  <section>
    <!---->
    <rk-tabs
      class="flex w-full "
      role="listbox"
      tabindex="0"
    >
      <rk-tab>
        <!---->
        <!---->
         
        <p>
          one
        </p>
        <!---->
        <!---->
         
        <!---->
      </rk-tab>
      <rk-tab>
        <!---->
        <!---->
         
        <p>
          two
        </p>
        <!---->
        <!---->
         
        <!---->
      </rk-tab>
      <rk-tab>
        <!---->
        <!---->
         
        <p>
          three
        </p>
        <!---->
        <!---->
         
        <!---->
      </rk-tab>
      <!---->
       
      <!---->
    </rk-tabs>
    
    <!---->
     
    <field-layout
      class="vertical"
    >
      <!---->
    </field-layout>
  </section>
  
</div>
`;

exports[`Category.svelte > should render with extra props 1`] = `
<div>
  <section>
    <!---->
    <rk-tabs
      class="flex w-full  justify-center"
      role="listbox"
      tabindex="0"
    >
      <rk-tab>
        <!---->
        <!---->
         
        <p>
          one
        </p>
        <!---->
        <!---->
         
        <!---->
      </rk-tab>
      <rk-tab>
        <!---->
        <!---->
         
        <p>
          two
        </p>
        <!---->
        <!---->
         
        <!---->
      </rk-tab>
      <rk-tab>
        <!---->
        <!---->
         
        <p>
          three
        </p>
        <!---->
        <!---->
         
        <!---->
      </rk-tab>
      <!---->
       
      <!---->
    </rk-tabs>
    
    <!---->
     
    <field-layout
      class="vertical"
    >
      <!---->
    </field-layout>
  </section>
  
</div>
`;

exports[`Category.svelte > should render with extra props 2`] = `
<div>
  <section>
    <!---->
    <rk-tabs
      class="flex w-full justify-end"
      role="listbox"
      tabindex="0"
    >
      <rk-tab>
        <!---->
        <!---->
         
        <p>
          one
        </p>
        <!---->
        <!---->
         
        <!---->
      </rk-tab>
      <rk-tab>
        <!---->
        <!---->
         
        <p>
          two
        </p>
        <!---->
        <!---->
         
        <!---->
      </rk-tab>
      <rk-tab>
        <!---->
        <!---->
         
        <p>
          three
        </p>
        <!---->
        <!---->
         
        <!---->
      </rk-tab>
      <!---->
       
      <!---->
    </rk-tabs>
    
    <!---->
     
    <field-layout
      class="vertical"
    >
      <!---->
    </field-layout>
  </section>
  
</div>
`;
